pkg/operator/ceph/object/notification/obc_label_controller.go

```diff
@@ -0,0 +1,216 @@
+	// Make sure a CephCluster is present otherwise do nothing
+	cephCluster, isReadyToReconcile, cephClusterExists, reconcileResponse := opcontroller.IsReadyToReconcileNoName(
+		r.client,
+		r.context,
+		controllerName,
+	)
+	if !isReadyToReconcile {
+		// This handles the case where the Ceph Cluster is gone and we want to delete that CR
+		if !cephOBC.GetDeletionTimestamp().IsZero() && !cephClusterExists {
+			// Return and do not requeue. Successful deletion.
+			return reconcile.Result{}, nil
+		}
+		logger.Debug("Ceph cluster not yet present")
+		return reconcileResponse, nil
+	}
+	r.clusterSpec = &cephCluster.Spec
+
+	// DELETE: the CR was deleted
+	if !cephOBC.GetDeletionTimestamp().IsZero() {
+		logger.Debugf("ObjectBucketClaim: %q was deleted", cephOBC.Name)
+		// Return and do not requeue. Successful deletion.
+		return reconcile.Result{}, nil
+	}
+
+	// Populate clusterInfo during each reconcile
+	r.clusterInfo, _, _, err = mon.LoadClusterInfo(r.context, cephCluster.Namespace)
+	if err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "failed to populate cluster info")
+	}
+
+	// create the notification if needed
+	if err = r.updateCephBucketNotifications(cephOBC); err != nil {
+		return reconcile.Result{}, errors.Wrapf(err, "failed to create bucket notifications")
+	}
+
+	// Other label based reconciliations should come here
+
+	// Return and do not requeue
+	logger.Debugf("ObjectBucketClaim: %q done reconciling labels", cephOBC.Name)
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileOBCLabels) updateCephBucketNotifications(cephOBC *bktv1alpha1.ObjectBucketClaim) error {
+	// step 1: get the ObjectBucket and StorageClass
+	namespace := cephOBC.Namespace
+	cephOB := bktv1alpha1.ObjectBucket{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: cephOBC.Spec.ObjectBucketName}, &cephOB); err != nil {
+		return errors.Wrapf(err, "failed to retrieve ObjectBucket: %q", cephOBC.Spec.ObjectBucketName)
+	}
+	cephSC := storagev1.StorageClass{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: cephOBC.Spec.StorageClassName}, &cephSC); err != nil {
+		return errors.Wrapf(err, "failed to retrieve StorageClass: %q", cephOBC.Spec.StorageClassName)
+	}
+
+	// step 2: delete all existing notifications
+	p := Provisioner{
+		Client:      r.client,
+		Context:     r.context,
+		ClusterInfo: r.clusterInfo,
+		ClusterSpec: r.clusterSpec,
+	}
+	if err := p.CreateSession(
+		cephOB.Spec.AdditionalState["cephUser"],
+		types.NamespacedName{
+			Name:      cephSC.Parameters["objectStoreName"],
+			Namespace: cephSC.Parameters["objectStoreNamespace"],
+		}); err != nil {
+		return errors.Wrapf(err, "failed to create session for bucket notification. bucket: %q", cephOB.Name)
+	}
+	err := p.DeleteAll(&cephOB)
+	if err != nil {
+		return errors.Wrapf(err, "failed delete all bucket notification from bucket: %q", cephOB.Name)
+	}
```

Comment:
Is this going to cause anyone's workloads or data collection to miss any notification events?

Reply:
a better approach would be to fetch the list of notifications from the RGW and do the diff with the labels
(this would require another extension to the AWS SDK)

Reply:
Sounds like we are on the same page there. I think this is good for the first version of the PR. We can come back to fix this in a follow-up PR if you'd prefer. But it is a definite TODO in my mind for releasing this.